segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/BookKeeperLog.java

```diff
@@ -709,16 +728,30 @@ private void rollover() {
             Ledgers.close(oldLedger);
             log.debug("{}: Rollover: swapped ledger and metadata pointers (Old = {}, New = {}) and closed old ledger.",
                     this.traceObjectId, oldLedger.getId(), newLedger.getId());
-        } catch (Throwable ex) {
-            if (!ExceptionHelpers.mustRethrow(ex)) {
-                log.error("{}: Rollover failure; log may be unusable.", ex);
-            }
+        } finally {
+            // It's possible that we have writes in the queue that didn't get picked up because they exceeded the predicted
+            // ledger length. Invoke the Write Processor to execute them.
+            this.writeProcessor.runAsync();
```

Comment:
Please clarify one thing for me. If there is an exception thrown in the try block, then the point of running the write processor to pick up the remaining writes (if any) would be to fail them and make sure they are retried?

Reply:
The Write Processor first checks if the current ledger is closed. If rollover failed, then the current ledger is closed, and it will reinvoke the rollover processor to do its job again. You may argue that the `handleRolloverFailures` already retriggers it, which is correct, but the `SequentialAsyncProcessor` will only execute it once, so there's no harm in doing so. This is cleaner than having a `try-catch-finally` block with some code in `catch` to figure out what to do. The addition of error handling in the SequentialAsyncProcessor eliminated this need by it itself tracking errors and invoking the handler.